core/metrics-spi/src/main/java/software/amazon/awssdk/metrics/SdkMetric.java

```diff
@@ -0,0 +1,129 @@
+
+    private final String name;
+    private final Class<T> clzz;
+    private final Set<MetricCategory> categories;
+
+    private SdkMetric(String name, Class<T> clzz, Set<MetricCategory> categories) {
+        this.name = Validate.notBlank(name, "name must not be blank");
+        this.clzz = Validate.notNull(clzz, "clzz must not be null");
+        this.categories = Validate.notEmpty(EnumSet.copyOf(categories), "categories must not be empty");
+    }
+
+    /**
+     * @return The name of this event.
+     */
+    public String name() {
+        return name;
+    }
+
+    /**
+     * @return The categories of this event.
+     */
+    public Set<MetricCategory> categories() {
+        return Collections.unmodifiableSet(categories);
+    }
+
+    /**
+     * @return The class of the value associated with this event.
+     */
+    public Class<T> valueClass() {
+        return clzz;
+    }
+
+    /**
+     * Cast the given object to the value class associated with this event.
+     *
+     * @param o The object.
+     * @return The cast object.
+     */
+    public T convertToType(Object o) {
+        return clzz.cast(o);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+
+        SdkMetric<?> sdkMetric = (SdkMetric<?>) o;
+
+        if (!name.equals(sdkMetric.name)) return false;
+        return clzz.equals(sdkMetric.clzz);
+    }
+
+    @Override
+    public int hashCode() {
+        int result = name.hashCode();
+        result = 31 * result + clzz.hashCode();
+        return result;
+    }
+
+    public static <T> SdkMetric<T> of(String name, Class<T> clzz, MetricCategory c1, MetricCategory... cn) {
+        return of(name, clzz, Stream.concat(Stream.of(c1), Stream.of(cn)).collect(Collectors.toSet()));
+    }
+
+        /**
+         * Create a new metric.
+         *
+         * @param name The name of this metric.
+         * @param clzz The class of the object containing the associated value for this metric.
+         * @param categories The categories associated with this metric.
+         * @param <T> The type of the object containing the associated value for this metric.
+         * @return The created metric.
+         *
+         * @throws IllegalArgumentException If a metric of the same name has already been created.
+         */
+    public static <T> SdkMetric<T> of(String name, Class<T> clzz, Set<MetricCategory> categories) {
+        Validate.noNullElements(categories, "categories must not contain null elements");
+        SdkMetric<T> event = new SdkMetric<>(name, clzz, categories);
+        synchronized (SDK_METRICS) {
+            if (SDK_METRICS.contains(event)) {
+                throw new IllegalArgumentException("Metric with name " + name + " has already been created");
+            }
+            SDK_METRICS.add(event);
+        }
```

Comment:
Have we considered pros and cons of using ConcurrentSkipListSet here instead of synchronizing?

Reply:
Or `ConcurrentHashMap.newKeySet()`

Reply:
Will take a look into alternative solutions that can take advantage of compare-and-set.